### DISCUSSION TITLE
How the partial derivatives R/w, f/w, R/x and f/x are computed?

### DISCUSSION THREAD

**Discussion by usmanshahid90** [mdolab/dafoam#296] [General]
Hello everyone, 
                         I have a very basic question in which I have stranded. we know that for finding the partial derivatives or simple derivatives, we need to have two initial and final values of the parameters involved. For an instance, if we need to find out the partial derivatives of R/w, f/w, R/x and f/x by using finite difference method then we need to have R1, R2, w1,w2 , x1, x2, f1 and f2. (f is objective function). 
As we are computing the flow only once, and are not even perturbing the design variables x then how we gonna compute the partial derivatives i.e. partial f/x. , partial R/x.
Moreover, as flow is computed only once then how flow variables are varying so that we could find the partial derivatives R/w (change in R with respect to change in w). and so on. 
Please guide me about it,

Thanks.

**Comment by chairmanmao256** [comment]
To the best of my knowledge, the partial derivatives (the Jacobians) are computed by automatic differentiation. The maintainer of this repository (friedenhe) wrote an article about the algorithm implemented in DAFoam:
He, Ping, et al. "An aerodynamic design optimization framework using a discrete adjoint approach with OpenFOAM." Computers & Fluids 168 (2018): 285-303.
I think this paper might help you to understand how the Jacobians are computed in DAFoam.
If you want further information on the underlying principles of automatic differentiation used, I suggest you read this book (I learn Automatic differentiation by reading this book and I think it is great):
Griewank, Andreas, and Andrea Walther. Evaluating derivatives: principles and techniques of algorithmic differentiation. Society for industrial and applied mathematics, 2008.
And [this package](https://github.com/SciCompKL/CoDiPack) might help you to obtain the derivatives of your own programs. And to my knowledge, DAFoam uses this package to derive the Jacobians.
